### PR TITLE
Add YYYYMMDD[hhmmss]to[Date, Date32, DateTime, DateTime64]()

### PR DESCRIPTION
**Pull request by rschu1ze**
Continues #45070, see #36470. Kudos to @dentiscalprum who build the first implementation - thanks a lot!

Decided to make a new PR as I rewrote major parts (the surrounding code changed a lot in the meantime) and @dentiscalprum is rather inactive in GitHub, he seems to be a student.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added functions YYYYMMDDtoDate(), YYYYMMDDtoDate32(), YYYYMMDDhhmmssToDateTime() and YYYYMMDDhhmmssToDateTime64(). They convert a date or date with time encoded as integer (e.g. 20230911) into a native date or date with time. As such, they provide the opposite functionality of existing functions YYYYMMDDToDate(), YYYYMMDDToDateTime(), YYYYMMDDhhmmddToDateTime(), YYYYMMDDhhmmddToDateTime64().

### CODE REVIEW

File: tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference [hunk 1]
```diff
@@ -56,6 +56,10 @@ URLHierarchy
 URLPathHierarchy
 UUIDNumToString
 UUIDStringToNum
+YYYYMMDDToDate
+YYYYMMDDToDate32
+YYYYMMDDhhmmssToDateTime
+YYYYMMDDhhmmssToDateTime64
```
Comment (Avogar) [inline tests/queries/0_stateless/02415_all_new_functions_must_be_documented.reference:62]: I don't think we should edit this file, as the name says, we should add documentation for all new functions in the code using `FunctionDocumentation argument in factory.registerFunction` . Let's do it for these new functions

File: src/Common/DateLUTImpl.h [hunk 2]
```diff
@@ -71,14 +71,14 @@ class DateLUTImpl
     // Same as above but select different function overloads for zero saturation.
     STRONG_TYPEDEF(UInt32, LUTIndexWithSaturation)
 
-    static inline LUTIndex normalizeLUTIndex(UInt32 index)
```
Comment (Avogar) [inline src/Common/DateLUTImpl.h:74]: What is the motivation of removing `inline` from this file?

Comment (rschu1ze) [reply src/Common/DateLUTImpl.h:74]: Functions defined inside class definitions are implicitly inline ([SO](https://stackoverflow.com/questions/9192077/is-inline-implicit-in-c-member-functions-defined-in-class-definition)). I find `inline` keywords slightly distracting in such contexts, so I decided to remove it.

**Comment by Avogar** [comment]
The code is perfect, I am ready to approve, but need to add at least minimal documentation to the code

**Comment by rschu1ze** [comment]
Merci. Will add minimal docs, of course.

(Just FYI: We are currently in the bad situation that docs are duplicated in markdown and as in-source docs. In an ideal world, the former is generated from the latter but we are not there yet (I think the docs team looks into this). So I am usually reluctant to add in-source docs as 1. they will potentially get out-of-sync with md docs and 2. as md docs are more complete than in-source docs, someone will need to update in-source docs when the tooling for auto-generating md exists.)

**Review by Avogar — APPROVED** [review]

**Comment by rschu1ze** [comment]
Let's save some CI cycles and merge, all tests were green in the previous iteration and only minor doc fixes were added.

**Pull request merged**